### ISSUE TITLE
Suppress one statement per line rule

### ISSUE THREAD

**Issue by maelstraggiotti** [Riverside-Software/sonar-openedge#330]
Hi,

At several places in our code base, we using a preprocessor directive to include some code that should be on separate lines (I didn't figured out how to include line breaks). 

Therefore, the "one statement per line" rule applies and I wish suppress it for these specific cases.

1. I can't find suppressing example in the document, I guess it's ```{&_proparse_ prolint-nowarn(onestatementperline)}``` but it doesn't seems to work.

2. If I include the directive into the .i file (that's it, a prepro in a prepro, is it supported ?), proparse crashes : 
```Error during code parsing for Class.cls:82:9: .cls
org.prorefactor.refactor.RefactorException: Class.cls:82:9: unexpected token: CATCH
	at org.prorefactor.treeparser.ParseUnit.parse(ParseUnit.java:152) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.treeparser.ParseUnit.treeParser01(ParseUnit.java:179) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.sonar.plugins.openedge.sensor.OpenEdgeProparseSensor.parseMainFile(OpenEdgeProparseSensor.java:193) [sonar-openedge-plugin-2.0.4-SNAPSHOT.jar:na]
	at org.sonar.plugins.openedge.sensor.OpenEdgeProparseSensor.execute(OpenEdgeProparseSensor.java:140) [sonar-openedge-plugin-2.0.4-SNAPSHOT.jar:na]
	at org.sonarsource.sonarlint.core.analyzer.sensor.SensorWrapper.analyse(SensorWrapper.java:52) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.analyzer.sensor.SensorsExecutor.executeSensor(SensorsExecutor.java:75) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.analyzer.sensor.SensorsExecutor.execute(SensorsExecutor.java:63) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.analyzer.sensor.PhaseExecutor.execute(PhaseExecutor.java:44) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.container.analysis.AnalysisContainer.doAfterStart(AnalysisContainer.java:143) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.container.ComponentContainer.startComponents(ComponentContainer.java:125) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.container.ComponentContainer.execute(ComponentContainer.java:110) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.container.storage.StorageAnalyzer.analyze(StorageAnalyzer.java:72) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.container.storage.StorageContainer.analyze(StorageContainer.java:126) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.ConnectedSonarLintEngineImpl.analyze(ConnectedSonarLintEngineImpl.java:158) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarsource.sonarlint.core.ConnectedSonarLintEngineImpl.analyze(ConnectedSonarLintEngineImpl.java:145) [org.sonarsource.sonarlint.core.sonarlint-core_2.11.0.RSSW.jar:na]
	at org.sonarlint.eclipse.core.internal.server.Server.runAnalysis(Server.java:273) [org.sonarlint.eclipse.core_3.0.999.jar:na]
	at org.sonarlint.eclipse.core.internal.jobs.AnalyzeProjectJob.run(AnalyzeProjectJob.java:472) [org.sonarlint.eclipse.core_3.0.999.jar:na]
	at org.sonarlint.eclipse.core.internal.jobs.AnalyzeProjectJob$AnalysisThread.run(AnalyzeProjectJob.java:125) [org.sonarlint.eclipse.core_3.0.999.jar:na]
Caused by: antlr.NoViableAltException: unexpected token: CATCH
	at org.prorefactor.proparse.ProParser.expression_statement(ProParser.java:1821) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.blockorstate(ProParser.java:426) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.code_block(ProParser.java:453) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.methodstate(ProParser.java:10384) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.statement(ProParser.java:2332) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.blockorstate(ProParser.java:422) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.code_block(ProParser.java:453) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.classstate(ProParser.java:5681) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.statement(ProParser.java:2038) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.blockorstate(ProParser.java:422) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.proparse.ProParser.program(ProParser.java:154) ~[proparse-1.0.17-SNAPSHOT.jar:na]
	at org.prorefactor.treeparser.ParseUnit.parse(ParseUnit.java:141) ~[proparse-1.0.17-SNAPSHOT.jar:na]
```

The directive looks like this : 
```
&GLOBAL-DEFINE ErrCatch {&_proparse_ prolint-nowarn(onestatementperline)} ~
CATCH sysErr AS Progress.Lang.SysError: ~
  MESSAGE "..". ~
END. ~
```
Also, the rules is applied even if I specify the suppress directive before ours : 
```
{&_proparse_ prolint-nowarn(onestatementperline)}
{&ErrCatch}
```

Thanks

**Comment by jurjendijkstra** [comment]
Yeah what a coincidence, I was just looking at the same thing. We too have IF..THEN DO: statement1. statement2. statement3. END. in a global-defined preprocessor and I did not find a way to rewrite it such that the rule does not give a warning. At last I just disabled the rule alltogether, but feel sorry about that and would rather have something like {&_proparse_ prolint-nowarn(onestatementperline)}

**Comment by gquerret** [comment]
> If I include the directive into the .i file (that's it, a prepro in a prepro, is it supported ?), proparse crashes :

Prepro in a prepro works. Except for proparse directives... 😱 
Will be fixed

**Comment by gquerret** [comment]
Status update:
* Proparse directive `onestatementperline` will be added
* Proparse directives will be handled as hidden tokens , not as statements anymore, that would fix #224 
* Proparse directives won't have any effect in `&GLOB-DEF` or `&SCOP-DEF`
* If multiple statements are found at the same, and **any** statement has a proparse directive, then the rule won't be triggered. That will allow this construct:
```
&GLOBAL-DEFINE ErrCatch CATCH sysErr AS Progress.Lang.SysError: MESSAGE "..". END.
{&_proparse_ prolint-nowarn(onestatementperline)}
{&ErrCatch}
```
or this one:
```
&GLOBAL-DEFINE XXX IF TRUE THEN DO: MESSAGE "xx". MESSAGE "zz". END.
{&_proparse_ prolint-nowarn(onestatementperline)}
{&XXX}
```

**Comment by gquerret** [comment]
And forgot to mention that it will be possible to use annotations in the future, i.e.
```
&GLOBAL-DEFINE ErrCatch @SuppressWarnings(type = "onestatementperline"). CATCH sysErr AS Progress.Lang.SysError: MESSAGE "..". END.
{&ErrCatch}
```

**Comment by gquerret** [comment]
Merged in master

**Comment by gquerret** [comment]
Problem is now fixed, see issue #877
